bot/bot.py

```diff
@@ -96,6 +112,14 @@ def _recreate(self) -> None:
                 "The previous connector was not closed; it will remain open and be overwritten"
             )
 
+        if self.redis_session and not self.redis_session.closed:
+            log.warning(
+                "The previous redis pool was not closed; it will remain open and be overwritten"
+            )
+
+        # Create the redis session
+        self.loop.create_task(self._create_redis_session())
```

Comment:
I'm not 100% certain about the usage of a bare `create_task()` here for creating the redis session. To me, it seems like you would want to guarantee that the session is created before proceeding rather than scheduling the initialization to be ran in the background whenever it has the chance to (which is usually while other coroutines are being awaited, but there's no guarantee that it will have time to complete).

In order to guarantee its completion, you could do one of the following, depending on whether or not blocking the event loop is a concern:

If blocking the event loop isn't a concern here, you could use `self.loop.run_until_complete(self._create_redis_session())`. If blocking the event loop is a concern, this function could be converted into a coroutine and then you could use `await self._create_redis_session()`. At a glance, it seems like blocking the event loop wouldn't be a significant issue here since it presumably would only be ran once on restart/setup.

Reply:
We've been discussing on Discord about reworking the creation of all the objects here. Creating events for each thing is just too cumbersome. Your proposal to block the loop is quite a novel approach to me. Somehow, it never crossed my mind to do it like this.

I'm inclined to leave this as is and rework  it along with creation of the other objects in a separate PR. Thoughts on this course of action?

Reply:
> Your proposal to block the loop is quite a novel approach to me. Somehow, it never crossed my mind to do it like this.

Yeah, I personally think that for initialization purposes when startup/restart is highly infrequent and doesn't take especially long, it's perfectly okay for some calls to block the event loop. This can help simplify the startup process significantly, and the performance impact is usually negligible at best.

> I'm inclined to leave this as is and rework it along with creation of the other objects in a separate PR. Thoughts on this course of action?

I think that would be reasonably okay. I would keep an eye on this part specifically to see if the creation of the redis session is ever not completed when expecting it to, although it looks like this is tracked by `self._redis_ready`. 

Also, a point in favor of converting to a coroutine that I forgot to mention is that it's technically "unsafe" to create async objects outside of a coroutine, especially without passing the event loop explicitly. In practice, this is only of particular significance when using an event loop that's not in the main thread or in general when using multiple event loops, but there's a decent chance that aiohttp will deprecate creating async objects outside of a coroutine and without passing the event loop, we've been doing this recently in the latest development versions of asyncio (such as for `asyncio.Lock()`).

It's caused enough interoperability issues between async libraries that it's being addressed now, although I suspect that a full removal would be decently far out due to the volume of possible breakage.

Reply:
Okay. This is an interesting bit of feedback, and thanks for that. I think I agree with what Mark is suggesting - that this should be refactored along with the other similar objects in a separate PR - so I'm going to resolve this conversation for now.

Reply:
@aeros the fact that it's not a coroutine is cause it's following discord.py's design. If a non-coroutine is called by a coroutine, then I think it's fine (which is the case when `login()` calls this). But there's also `clear()`, which isn't a coroutine. And you can see I noted in a comment that calling it outside a coroutine will raise a warning. Discord.py creates its session outside a coroutine and supports re-creating it by calling `clear()` and I'm just going along with this design though we were discussing letting it go.